tests/meshcop/commissioner.cpp

```diff
@@ -28,333 +28,305 @@
+    mbedtls_entropy_init(&mEntropy);
+    mDtlsInitDone = true;
+    SuccessOrExit(ret = mbedtls_ctr_drbg_seed(&mDrbg, mbedtls_entropy_func, &mEntropy, kSeed, sizeof(kSeed)));
+    SuccessOrExit(ret = mbedtls_net_connect(&mSslClientFd, addressAscii, portAscii, MBEDTLS_NET_PROTO_UDP));
+    SuccessOrExit(ret = mbedtls_ssl_config_defaults(&mSslConf, MBEDTLS_SSL_IS_CLIENT, MBEDTLS_SSL_TRANSPORT_DATAGRAM,
+                                                    MBEDTLS_SSL_PRESET_DEFAULT));
+    mbedtls_ssl_conf_rng(&mSslConf, mbedtls_ctr_drbg_random, &mDrbg);
+    mbedtls_ssl_conf_min_version(&mSslConf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);
+    mbedtls_ssl_conf_max_version(&mSslConf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);
+    mbedtls_ssl_conf_authmode(&mSslConf, MBEDTLS_SSL_VERIFY_NONE);
+    mbedtls_ssl_conf_dbg(&mSslConf, MBedDebugPrint, stdout);
+    mbedtls_ssl_conf_ciphersuites(&mSslConf, kCipherSuites);
+    mbedtls_ssl_conf_export_keys_cb(&mSslConf, DummyKeyExport, NULL);
+    mbedtls_ssl_conf_handshake_timeout(&mSslConf, kMbedDtlsHandshakeMinTimeout, kMbedDtlsHandshakeMaxTimeout);
 
-    addrlen = sizeof(from_addr);
+    otbrLog(OTBR_LOG_INFO, "connecting: ssl-setup");
+    SuccessOrExit(ret = mbedtls_ssl_setup(&mSsl, &mSslConf));
+    mbedtls_ssl_set_bio(&mSsl, &mSslClientFd, mbedtls_net_send, mbedtls_net_recv, mbedtls_net_recv_timeout);
+    mbedtls_ssl_set_timer_cb(&mSsl, &mTimer, mbedtls_timing_set_delay, mbedtls_timing_get_delay);
 
-    ssize_t ret = recvfrom(aContext.mSocket, dtlsEncapsulation, sizeof(dtlsEncapsulation), 0,
-                           (struct sockaddr *)(&from_addr), &addrlen);
+    mbedtls_ssl_set_hs_ecjpake_password(&mSsl, mPskcBin, OT_PSKC_LENGTH);
 
-    VerifyOrExit(ret > 0);
+    otbrLog(OTBR_LOG_INFO, "connect: perform handshake");
+    do
     {
-        // Print this, because in some environments...
-        // other things on the network use the same port
-        // as open thread is using here :-(
-        char buf[IPSTR_BUFSIZE];
-        get_ip_str((struct sockaddr *)(&from_addr), buf, sizeof(buf));
-        otbrLog(OTBR_LOG_INFO, "relay from: %s\n", buf);
-    }
-
-    responseTlv->SetType(Meshcop::kJoinerDtlsEncapsulation);
-    responseTlv->SetValue(dtlsEncapsulation, static_cast<uint16_t>(ret));
-    responseTlv = responseTlv->GetNext();
-
-    responseTlv->SetType(Meshcop::kJoinerUdpPort);
-    responseTlv->SetValue(aContext.mJoiner.mUdpPort);
-    responseTlv = responseTlv->GetNext();
-
-    responseTlv->SetType(Meshcop::kJoinerIid);
-    responseTlv->SetValue(aContext.mJoiner.mIid, sizeof(aContext.mJoiner.mIid));
-    responseTlv = responseTlv->GetNext();
-
-    responseTlv->SetType(Meshcop::kJoinerRouterLocator);
-    responseTlv->SetValue(aContext.mJoiner.mRouterLocator);
-    responseTlv = responseTlv->GetNext();
+        ret = mbedtls_ssl_handshake(&mSsl);
+    } while (ret == MBEDTLS_ERR_SSL_WANT_READ || ret == MBEDTLS_ERR_SSL_WANT_WRITE);
```

Comment:
This effectively blocks until the handshake completes. We should consider having this class support single-threaded processing via a main-loop.